cbmc/proofs/ParseDNSReply/ParseDNSReply_harness.c

```diff
@@ -0,0 +1,165 @@
+/* Standard includes. */
+#include <stdint.h>
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+#include "queue.h"
+#include "list.h"
+#include "semphr.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_Sockets.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_UDP_IP.h"
+#include "FreeRTOS_DNS.h"
+#include "NetworkBufferManagement.h"
+#include "NetworkInterface.h"
+#include "IPTraceMacroDefaults.h"
+
+#include "cbmc.h"
+
+uint32_t prvParseDNSReply(uint8_t *pucUDPPayloadBuffer,
+			  size_t xBufferLength,
+			  TickType_t xIdentifier);
+
+// Used in specification and abstraction of prvReadNameField, prvSkipNameField
+// Given unconstrained values in harness
+uint8_t *buffer;
+size_t buffer_size;
+
+// Abstraction of prvReadNameField
+uint8_t *prvReadNameField(uint8_t *pucByte,
+			  size_t xSourceLen,
+			  char *pcName,
+			  size_t xDestLen){
+
+  // Preconditions
+
+  // pointers are valid pointers into buffers
+  __CPROVER_assert(xSourceLen == 0 ||
+		   (buffer <= pucByte && pucByte < buffer + buffer_size),
+		   "pucByte valid pointer");
+  //  __CPROVER_assume(name <= pcName && pcName < name + name_size);
+  (void) *(pcName); // can only test for valid pointer
+
+  // lengths are valid values for space remaining in the buffers
+  __CPROVER_assert(pucByte + xSourceLen <= buffer + buffer_size,
+		   "xSourceLen valid length");
+  //  __CPROVER_assume(pcName + xDestLen <= name + name_size);
+  (void) *(pcName + (xDestLen-1)); // can only test for valid pointer
+
+  // CBMC loop unwinding: bounds depend on xSourceLen and xDestLen
+  __CPROVER_assert(xSourceLen <= NETWORK_BUFFER_SIZE,
+		   "xSourceLen loop unwinding");
+  //  __CPROVER_assume(xDestLen <= NAME_SIZE);
+
+  // Buffer overflow from integer overflow in expression len < xDestLen-1
+  // xDestLen == 254 in code
+  __CPROVER_assert(xDestLen > 0, "xDestLen nonzero");
+
+  SAVE_OLD(pucByte, uint8_t *);
```

Comment:
There are scattered macro calls that are not accompanied by inline comments, although I think they would benefit from them.